File: behavioral/visitor.ts
// Visitor Design Pattern Overview
// The Visitor pattern allows you to add operations to classes without having to modify them.
// It achieves this by separating the structure (the elements being operated upon) from the operations (the visitor).

// Example Scenario
// Imagine a scenario where you have different types of products in a shopping cart,
// and you want to calculate the shipping cost for each type of product differently.
// This is where the Visitor pattern can be applied: each product can accept a visitor that calculates its shipping cost based on its type.

// Visitor Interface
interface ProductVisitor {
  visitBook(product: Book): number;
  visitElectronics(product: Electronics): number;
}

// Concrete Visitor: ShippingCostCalculator
class ShippingCostCalculator implements ProductVisitor {
  visitBook(product: Book): number {
    // Calculate shipping cost for a book
    return product.weight * 0.5; // Example: $0.5 per gram
  }

  visitElectronics(product: Electronics): number {
    // Calculate shipping cost for electronics
    return product.weight * 1.2; // Example: $1.2 per gram
  }
}

// Element Interface
interface Product {
  accept(visitor: ProductVisitor): number;
}

// Concrete Elements
class Book implements Product {
  constructor(public weight: number) {}

  accept(visitor: ProductVisitor): number {
    return visitor.visitBook(this);
  }
}

class Electronics implements Product {
  constructor(public weight: number) {}

  accept(visitor: ProductVisitor): number {
    return visitor.visitElectronics(this);
  }
}

// Example Usage
const book = new Book(500); // 500 grams
const electronics = new Electronics(2000); // 2000 grams

const shippingCostCalculator = new ShippingCostCalculator();

const bookShippingCost = book.accept(shippingCostCalculator);
console.log(`Shipping cost for the book: $${bookShippingCost}`);

const electronicsShippingCost = electronics.accept(shippingCostCalculator);
console.log(`Shipping cost for electronics: $${electronicsShippingCost}`);

File: behavioral/visitor.py
from abc import ABC, abstractmethod


# Visitor Interface
class ProductVisitor(ABC):
    @abstractmethod
    def visit_book(self, product: "Book") -> float:
        pass

    @abstractmethod
    def visit_electronics(self, product: "Electronics") -> float:
        pass


# Concrete Visitor: ShippingCostCalculator
class ShippingCostCalculator(ProductVisitor):
    def visit_book(self, product: "Book") -> float:
        # Calculate shipping cost for a book
        return product.weight * 0.5  # Example: $0.5 per gram

    def visit_electronics(self, product: "Electronics") -> float:
        # Calculate shipping cost for electronics
        return product.weight * 1.2  # Example: $1.2 per gram


# Element Interface
class Product(ABC):
    @abstractmethod
    def accept(self, visitor: ProductVisitor) -> float:
        pass


# Concrete Elements
class Book(Product):
    def __init__(self, weight: int):
        self.weight = weight

    def accept(self, visitor: ProductVisitor) -> float:
        return visitor.visit_book(self)


class Electronics(Product):
    def __init__(self, weight: int):
        self.weight = weight

    def accept(self, visitor: ProductVisitor) -> float:
        return visitor.visit_electronics(self)


# Example Usage
book = Book(500)  # 500 grams
electronics = Electronics(2000)  # 2000 grams

shipping_cost_calculator = ShippingCostCalculator()

book_shipping_cost = book.accept(shipping_cost_calculator)
print(f"Shipping cost for the book: ${book_shipping_cost}")

electronics_shipping_cost = electronics.accept(shipping_cost_calculator)
print(f"Shipping cost for electronics: ${electronics_shipping_cost}")
